src/index/CompressedRelationPermutationWriterImpl.h

```diff
@@ -0,0 +1,288 @@
+      twinRelationSorter_;
+
+  DistinctIdCounter distinctCol1Counter_;
+
+  using PerBlockCallbacks =
+      std::vector<std::function<void(const IdTableStatic<0>&)>>;
+  const PerBlockCallbacks perBlockCallbacks_;
+
+  // Set up the handling of small relations for the twin permutation.
+  // A complete block of them is handed from `writer1` to the following lambda
+  // (via the `smallBlocksCallback_` mechanism. The lambda the resorts the
+  // block and feeds it to `writer2`.)
+  struct AddBlockOfSmallRelationsToSwitched {
+    CompressedRelationWriter& writer_;
+    void operator()(IdTable relation) const {
+      // We don't use the parallel twinRelationSorter to create the twin
+      // relation as its overhead is far too high for small relations.
+      relation.swapColumns(c1Idx_, c2Idx_);
+
+      // We only need to sort by the columns of the triple + the graph
+      // column, not the additional payload. Note: We could also use
+      // `compareWithoutLocalVocab` to compare the IDs cheaper, but this
+      // sort is far from being a performance bottleneck.
+      auto compare = [](const auto& a, const auto& b) {
+        return std::tie(a[0], a[1], a[2], a[3]) <
+               std::tie(b[0], b[1], b[2], b[3]);
+      };
+      ql::ranges::sort(relation, compare);
+      AD_CORRECTNESS_CHECK(!relation.empty());
+      // Note: it is important that we store these two IDs before moving the
+      // `relation`, because the evaluation order of function arguments is
+      // unspecified.
+      auto firstCol0 = relation.at(0, 0);
+      auto lastCol0 = relation.at(relation.numRows() - 1, 0);
+      writer_.compressAndWriteBlock(firstCol0, lastCol0, std::move(relation),
+                                    false);
+    };
+  };
+
+  // ___________________________________________________________________________
+  PermutationWriter(const std::string& basename,
+                    WriterAndCallback writerAndCallback1,
+                    WriterAndCallback writerAndCallback2,
+                    qlever::KeyOrder permutation,
+                    PerBlockCallbacks perBlockCallbacks)
+      : permutation_{std::move(permutation)},
+        writer1_{writerAndCallback1.writer_},
+        writer2_{writerAndCallback2.writer_},
+        writeMetadata_{std::move(writerAndCallback1.callback_),
+                       std::move(writerAndCallback2.callback_),
+                       writerAndCallback1.writer_.blocksize()},
+        blocksize_{writerAndCallback1.writer_.blocksize()},
+        numColumns_{writerAndCallback1.writer_.numColumns()},
+        relation_{numColumns_, alloc_},
+        twinRelationSorter_{basename + ".twin-twinRelationSorter",
+                            numColumns_,
+                            4_GB,
+                            alloc_,
+                            ad_utility::DEFAULT_BLOCKSIZE_EXTERNAL_ID_TABLE,
+                            TwinComparator{}},
```

Comment:
```suggestion
                            },
```
This was only needed because lambdas in C++17 are not default-constructible, but now you have a proper struct again:)